### PR TITLE
Provide a Plugin#getOSGiService method

### PR DESCRIPTION
**Pull request by laeubi**
https://github.com/eclipse-platform/eclipse.platform.runtime/issues/27

### CODE REVIEW

File: bundles/org.eclipse.core.runtime/src/org/eclipse/core/runtime/Plugin.java [hunk 1]
```diff
@@ -526,6 +527,35 @@ public void setDebugging(boolean value) {
...
+	/**
+	 * Try to acquire a single service from the OSGi service factory (e.g.
+	 * DebugOptions, IWorkspace, IPreferenceService, ...)
+	 *
+	 * @param service the service to acquire
+	 * @return an optional representing the acquired service or an empty optional if
+	 *         there is (currently) no service available or the plugin is shut down
+	 * @since 3.25
+	 */
+	public <T> Optional<T> getOSGiService(Class<T> service) {
```
Comment (iloveeclipse) [inline bundles/org.eclipse.core.runtime/src/org/eclipse/core/runtime/Plugin.java:539]: "service" should be better "serviceClass", otherwise it sounds strange to ask this method for a service giving service as argument.

File: bundles/org.eclipse.core.runtime/src/org/eclipse/core/runtime/Plugin.java [hunk 2]
```diff
@@ -526,6 +527,35 @@ public void setDebugging(boolean value) {
 		}
 	}
 
+	/**
+	 * Try to acquire a single service from the OSGi service factory (e.g.
+	 * DebugOptions, IWorkspace, IPreferenceService, ...)
+	 *
+	 * @param service the service to acquire
```
Comment (iloveeclipse) [inline bundles/org.eclipse.core.runtime/src/org/eclipse/core/runtime/Plugin.java:534]: Class of the service to acquire

**Comment by iloveeclipse** [comment]
Would be also nice to have tests for new API added

File: bundles/org.eclipse.core.runtime/src/org/eclipse/core/runtime/Plugin.java (hunk 1 above)
Comment (mickaelistria) [inline bundles/org.eclipse.core.runtime/src/org/eclipse/core/runtime/Plugin.java:539]: what about naming it just `lookup()` as it's done in many other containers?

File: bundles/org.eclipse.core.runtime/src/org/eclipse/core/runtime/Plugin.java [hunk 3]
```diff
@@ -526,6 +527,35 @@ public void setDebugging(boolean value) {
 		}
 	}
 
+	/**
+	 * Try to acquire a single service from the OSGi service factory (e.g.
+	 * DebugOptions, IWorkspace, IPreferenceService, ...)
```
Comment (mickaelistria) [inline bundles/org.eclipse.core.runtime/src/org/eclipse/core/runtime/Plugin.java:532]: Would it be worth adding an @implNote telling it relies on the ServiceTracker, so consumer can get a sense of how they can then inject their own service?

Comment (laeubi) [reply bundles/org.eclipse.core.runtime/src/org/eclipse/core/runtime/Plugin.java:532]: I don't think this is specific to how the service is acquired and users can't learn much about if the know how it works internally. "OSGi service factory" actually tells where service are aquired and how to provide own implementations.

File: bundles/org.eclipse.core.runtime/src/org/eclipse/core/runtime/Plugin.java (hunk 1 above)
Comment (laeubi) [reply bundles/org.eclipse.core.runtime/src/org/eclipse/core/runtime/Plugin.java]: done

Comment (laeubi) [reply bundles/org.eclipse.core.runtime/src/org/eclipse/core/runtime/Plugin.java]: done

File: bundles/org.eclipse.core.runtime/src/org/eclipse/core/runtime/Plugin.java (hunk 2 above)
Comment (laeubi) [reply bundles/org.eclipse.core.runtime/src/org/eclipse/core/runtime/Plugin.java]: done

**Comment by laeubi** [comment]
> Would be also nice to have tests for new API added

I think the `DebugOption` test should cover this already as I replaced the explicit tracer for those with the new method.

**Comment by laeubi** [comment]
> > Would be also nice to have tests for new API added
> 
> I think the `DebugOption` test should cover this already as I replaced the explicit tracer for those with the new method.

Okay it seems there is no test referencing this, I'll add a dedicated one...

**Comment by laeubi** [comment]
Test is added now.

**Comment by tjwatson** [comment]
Use of the `org.eclipse.core.runtime` BundleContext object to get services on behalf of other bundles is not a best practice.  We should not add such APIs to the platform.

File: bundles/org.eclipse.core.runtime/src/org/eclipse/core/runtime/Plugin.java [hunk 4]
```diff
@@ -526,6 +525,63 @@ public void setDebugging(boolean value) {
...
+	 *
+	 * @param serviceType type of the service to acquire
+	 * @return an optional representing the acquired service or an empty optional if
+	 *         there is (currently) no service available or the plugin is shut down
+	 * @since 3.25
+	 */
+	public <T> Optional<T> lookup(Class<T> serviceType) {
```
Comment (tjwatson) [inline bundles/org.eclipse.core.runtime/src/org/eclipse/core/runtime/Plugin.java:537]: For this to work you need to access to the callers `BundleContext` then you need to track and get the service using that context, not the core.runtime context.  Then you need to track that bundle and react when the bundle is stopped so that you flush the usage of the service properly.

If you don't do this then you mess with the logic of the ServiceFactory that may depend on known what bundle is using its service.

Please have a look at `org.eclipse.core.runtime.ServiceCaller<Service>` and consider enhancing it with some of these ideas.  For example, adding a CompletableFuture.

**Review by tjwatson — CHANGES_REQUESTED** [review]

**Comment by tjwatson** [comment]
Just to illustrate the usage of ServiceCaller from the javadoc:

```java
 static final ServiceCaller<ILog> log = new ServiceCaller(MyClass.class, ILog.class);
 static void info(String msg) {
   log.call(logger -> logger.info(msg));
 }
```
This does all the things I mention by using the bundle `MyClass` is in to get its `BundleContext` to track and obtain the `ILog` service.  The `call` method then conditionally calls the passed in function depending on if the service is available.

File: bundles/org.eclipse.core.runtime/src/org/eclipse/core/runtime/Plugin.java (hunk 4 above)
Comment (laeubi) [reply bundles/org.eclipse.core.runtime/src/org/eclipse/core/runtime/Plugin.java:537]: @tjwatson this does use the context of the plugin bundle not the` org.eclipse.core.runtime` BundleContext ...

Comment (tjwatson) [reply bundles/org.eclipse.core.runtime/src/org/eclipse/core/runtime/Plugin.java:537]: Ah, right!  Sorry I was thinking this was the `Platform` class.  Regardless I think the `ServiceCaller` should be considered instead of adding more API for accessing services.

Comment (laeubi) [reply bundles/org.eclipse.core.runtime/src/org/eclipse/core/runtime/Plugin.java:537]: No problem, do you like to remove your change requesting review then?
`ServiceCaller` replicates unnecessary many code of of `ServiceTracker` so I never got why I should use this over the other.

Comment (tjwatson) [reply bundles/org.eclipse.core.runtime/src/org/eclipse/core/runtime/Plugin.java:537]: It depends on what that means to remove my change request review.  I am not convinced adding this new API to `Plugin` is necessary given that I think `ServiceCaller`can be used in most cases and we can consider enhancing it if we think some kind of future capability is needed.

I'm not sure what you mean by replicates code of `ServiceTracker`.  The advantage of `ServiceCaller` vs. `ServiceTracker` is simplicity of use over `ServiceTracker` and that it allows for `single` use service calls, as well as support for long-term usage of a service with many calls to the service.  The javadoc for `ServiceCaller` claims the following about `ServiceTracker`:

```
does not easily allow for lazy instantiation of the service instance.
```
Although I don't recall why that statement is true.  Perhaps @alblue can recall why he thought `ServiceTracker` made it harder to do lazy instantiation.

Comment (laeubi) [reply bundles/org.eclipse.core.runtime/src/org/eclipse/core/runtime/Plugin.java:537]: > does not easily allow for lazy instantiation of the service instance.

Yep, I have had a discussion  with @alblue a while back, this claim is simply wrong and if one really wan't to have "call once" or similar all this could be done by delegating to a lazy fetching service tracker (I used one for example when implementing the AdapterFactory allowing OSGi services).

Comment (laeubi) [reply bundles/org.eclipse.core.runtime/src/org/eclipse/core/runtime/Plugin.java:537]: > It depends on what that means to remove my change request review.

Currently you have requested a change on "For this to work you need to access to the callers BundleContext then you need to track and get the service using that context, not the core.runtime context.", you could resolve this conversation and simply do another review requesting an updated code location or comment.

**Review by tjwatson — CHANGES_REQUESTED** [review]
For the record, I prefer that we do not add more APIs for accessing the service registry.  I prefer we gravitate to using `ServiceCaller` instead.  If others chime in and think this new API is warranted then perhaps I can be convinced otherwise.

**Comment by laeubi** [comment]
> For the record, I prefer that we do not add more APIs for accessing the service registry.

I think this is valid and I think OSGi already offers anything we need, but still it was requested to have some convenient access (in contrast to using static singletons) in the very specific context of an Eclipse Plugin.

> I prefer we gravitate to using `ServiceCaller` instead. If others chime in and think this new API is warranted then perhaps I can be convinced otherwise.

I really have some strong concerns at recommend `ServiceCaller`usage:

1. It (unnecessarily) replicates a lot of code & care of [ServiceTracker](http://docs.osgi.org/specification/osgi.core/7.0.0/util.tracker.html), the OSGi spec itself recommend to not do that (emphasis by me):
"The dynamic nature of the Framework service registry makes it necessary to track the service objects as they are registered and unregistered to prevent problems. **It is easy to overlook race conditions or boundary conditions that will lead to random errors**. "
Looking at the code I'm not convinced it handles all these aspects there are only a few test and none of them seem to handle concurrency, and even if, why redo all the work already provided by the framework? 
2. Its claim that service tracker could not be used "lazy" is simply wrong and gives false impressions to users that service-tracker is bad (what it isn't) or even causing issues (whats wrong), its just a bit low-level (bus still very powerful).
3. As given by its name, its dedicated to calling a service method. While for simple cases (e.g. logging, firing an event) this is appropriate, I almost always have the more common case that I need to interact with a service (e.g. over a long time), probably need to fetch some return values and alike and in the case of registering a listener I need to make sure that I remove the listener from the *same* service object and so on. That results in that people use some techniques to "get the service out of the closure", e.g assigning it to fields or passing it to other objects. Sometimes just to have checked exceptions instead of runtime exceptions.
4. The "callonce" method might cause "flicker" lazy DS components if it get/unget a service but people will tend to use such methods as they seem very convenient (e.g. no need to hold a reference anywhere).

Beside that, for the context of the change all that effort finding out the BC, tracking bundles  is not necessary from within an(Plugin)Activator as it plugin already handles its life-cycle appropriately.

**Comment by tjwatson** [comment]
> 
> I really have some strong concerns at recommend `ServiceCaller`usage:

If the concerns are valid and not fixable then we should deprecate `ServiceCaller` before adding more API that provides yet another way to access services. Although, I would prefer to fix the API we got, if there are problems with it instead of adding more API.

> 
> 1. It (unnecessarily) replicates a lot of code & care of [ServiceTracker](http://docs.osgi.org/specification/osgi.core/7.0.0/util.tracker.html), the OSGi spec itself recommend to not do that (emphasis by me):
>    "The dynamic nature of the Framework service registry makes it necessary to track the service objects as they are registered and unregistered to prevent problems. **It is easy to overlook race conditions or boundary conditions that will lead to random errors**. "

But we again are trying to provide "a better way to get services" here.

For the record, there was a bit of discussion about `ServiceCaller` in https://bugs.eclipse.org/bugs/show_bug.cgi?id=563987

>    Looking at the code I'm not convinced it handles all these aspects there are only a few test and none of them seem to handle concurrency, and even if, why redo all the work already provided by the framework?

I do believe the `ServiceCaller` class does handle these aspects of the service registry.  If you see flaws we should fix them.  One thing `ServiceCaller` does not do that `ServiceTracker` does is track more than one service.  `ServiceCaller` only tracks the highest ranked service.  It also is not extensible like `ServiceTracker` so therefore is much less complicated (implementation wise) than `SerivceTracker`.  


> 2. Its claim that service tracker could not be used "lazy" is simply wrong and gives false impressions to users that service-tracker is bad (what it isn't) or even causing issues (whats wrong), its just a bit low-level (bus still very powerful).

Fair enough, but not sure that is a reason not to use `ServiceCaller`.  But we should remove that part of the `ServiceCaller` javadoc IMO.

> 3. As given by its name, its dedicated to calling a service method. While for simple cases (e.g. logging, firing an event) this is appropriate, I almost always have the more common case that I need to interact with a service (e.g. over a long time), probably need to fetch some return values and alike and in the case of registering a listener I need to make sure that I remove the listener from the _same_ service object and so on. That results in that people use some techniques to "get the service out of the closure", e.g assigning it to fields or passing it to other objects. Sometimes just to have checked exceptions instead of runtime exceptions.

I'm not sure how this solution solves that problem.  Calls to tracker.getService can return a different service object each call.  The suggested Optional in this API seems the same conceptually as `org.eclipse.core.runtime.ServiceCaller.current()`.  Sure `ServiceCaller` has other methods there for convenience in doing single method calls, but it can be used very easily for multiple method call usage.

> 4. The "callonce" method might cause "flicker" lazy DS components if it get/unget a service but people will tend to use such methods as they seem very convenient (e.g. no need to hold a reference anywhere).

Right, I don't suggest we push usage of `callOnce` widely.  There are very few cases where I think `callOnce` is appropriate to use.  But for the `Plugin` API to be used widely the `Plugin` instance is going to have to be passed around to other parts of the bundle to get access or the bundles using this will have to provide a static access to the Plugin instance (which is not good either).  On the other hand `ServiceCaller` is perfectly safe to assign in a static final variable.

> 
> Beside that, for the context of the change all that effort finding out the BC, tracking bundles is not necessary from within an(Plugin)Activator as it plugin already handles its life-cycle appropriately.

Sure, but I still do not advise adding even more API to access the low-level service registry.  I think we have enough abstracts on top of the service registry already.

**Comment by laeubi** [comment]
> Although, I would prefer to fix the API we got, if there are problems with it instead of adding more API.

I'm all for that if the ratio between effort of implementation and effort of "convincing" people is not completely off... and I'm really in fear it will end in an endless "why do we change everything is fine" discussion...

> If you see flaws we should fix them

As i generally avoid using ServiceCaller I can't point to any direct ones, but always found adapting classes to bundles and fetching then the BC is often flawed as it makes assumptions as how "good bundles" has to behave (e.g. lazy started) and in fact are often indicate that the code is not well aware of the fact that it is using OSGi, so yes using (static) Plugin Activators is not better per se, but at least it participates in the OSGi lifecycle without too much assumptions and restrictions.

> But we should remove that part of the `ServiceCaller` javadoc IMO.

I would really appreciate this, as it really scares me if I'm reading this ... but that might be more a personal feeling ;-)

> I'm not sure how this solution solves that problem.

This could also suffer from such problems (as the current static accessory does), but returning the service is just much more familiar for people migrating from static `Platform.getXY` / `AbcPlugin.getDefault().getZ() ` ... So don'T get me wrong I'd like to use that all over the place, I just try to solve a very specific request from people that claim the have no idea how they should replace such calls with a call into the service registry :-)

> I still do not advise adding even more API to access the low-level service registry. I think we have enough abstracts on top of the service registry already.

As I can't really judge it I'd like to get some feedback here:

@iloveeclipse @mickaelistria wdyt? Would `ServiceCaller` be the way to go (and advertise) for migration purpose?
@merks Would you feel comfortable in using a `ServiceCaller` API in favor of eg. ResourcePlugin.getWorkspace() does you feel anything missing or confusing?

It still might be good to back ServiceCaller by ServiceTracker (wich could be customized to current implementation) and migrate the track() method in favor of this PR... but as mentioned before I'd only like to to this if there is a minimal consent in doing so.

**Comment by mickaelistria** [comment]
I'm not as familiar with services as @laeubi  or @tjwatson , so I'm mostly learning here. I'm looked at ServiceCaller and it seems like a fine entry-point, and sympathize with @tjwatson that adding more entry-point for a functionality is usually not a good thing. So if we can be happy with ServiceCaller in other parts of Platform , let's just stick to it.

**Comment by alblue** [comment]
> > Its claim that service tracker could not be used "lazy" is simply wrong and gives false impressions to users that service-tracker is bad (what it isn't) or even causing issues (whats wrong), its just a bit low-level (bus still very powerful).
> Fair enough, but not sure that is a reason not to use ServiceCaller. But we should remove that part of the ServiceCaller javadoc IMO.

The problem is, the service tracker cannot be used lazily and therefore this statement is correct. See https://github.com/alblue/why-servicetracker-is-bad for more information. I know that @laeubi disagrees (we've had discussions on this before) but that doesn't change the fact that it is bad.

The specific problem is that to use a `ServiceTracker` can't be used to interact with services until the `.open()` call is invoked. However, when `.open()` is invoked it triggers a servlce lookup. This in turn is picked up by DS which then eagerly forces the activation of the service **during the `.open()` call synchronously** so that by the time the `.open()` call returns, it has activated all services which matched during the lookup from DS providers which otherwise would be lazily used on demand.

The advantage of the `ServiceCaller` is that it doesn't do this; rather, it defers the time it looks up the service until it is needed, and only then does it perform the lookup. So you can have a reference to a service (interface) and delay the instantiation of the component(a) until such time as you actually need to call them.

There is a further problem with using `ServiceTracker` in the use of static singleton variables (for which old-style Eclipse 2.x code is well known) in that you can't (easily) have a variable assigned to the component without opening it, but since the open call doesn't return the type, you can't use (e.g.) `MyService x = new ServiceTracker().open()`. Instead you have to worry about putting the `.open()` into a `static{}` block, and then deal with any errors that might occur during the call of the open itself.

The `ServiceCaller` was designed to be a drop in replacement for some of the platform-wide singletons with a more fluent/chainable API so that you could e.g. get/call the `ILog` service in a trivial way. Importantly, since this class also only depends on pure OSGi classes, there's no need to have it dependent upon (e.g.) Platform or Common; you can just use it in any class across Eclipse.

One of the benefits of getting rid of the `Activator` classes over the past few years has been to remove `Plugin` classes which are otherwise unnecessary, both to reduce code and to reduce coupling to the Eclipse Platform specific plugins.

Adding a reference to get this service to the `Plugin` class is therefore exactly the opposite of what many of the changes were meant to do, which is to couple it tighter to the `Plugin`. Instead this will encourage uses of code referring to the `Plugin` class even when they don't need to when better alternatives exist.

Finally, the use of the callback within the `ServiceCaller` was to prevent end users from capturing that specific service and storing it in a persistent location, so that if you needed to interact with Service X you had it (for the duration of the method call) and then you didn't have a reference to it afterwards, much like a try-with-resources would work. This allows you to cleanly execute a service but not require an instance to it, and as such, when a new service is published in place subsequent `callOnce` calls will then use that newly provisioned instance instead of capturing the old one. This method added appears to violate the dynamism of the OSGi API and to return a service which could be inadvertently cached in an instance variable, which would be incorrect.

**Comment by laeubi** [comment]
> The problem is, the service tracker cannot be used lazily and therefore this statement is correct.

Please take a look at https://github.com/eclipse-equinox/equinox.bundles/blob/master/bundles/org.eclipse.equinox.common/src/org/eclipse/core/internal/runtime/AdapterFactoryBridge.java is is lazy, even super-lazy because it delays the fetching of the service up to the point where actually a call is made to the adapter factory.

> See https://github.com/alblue/why-servicetracker-is-bad for more information

You just haven't shown that DS makes a difference here, if you have a DS component as your client it will also activate the provider service immediate, so if you comparing things you should compare the same things, no one forces you to create and open the tracker in the activator (even though that's convenient and often don't make a difference).

So the statement still does not holds true that `ServiceTracker` can't be used "lazy" and that DS is lazy by default, DS just handel most aspects in a more convenient way, but if you like to have a lazy DS component as in "do not lazily activate a bundle" then you have to use similar code as with service-tracker.

> that you can't (easily) have a variable assigned to the component without opening it,

Huh? Beside that it is always bad to uses that static pattern why should one not assign it to a static variable?

> but since the open call doesn't return the type, you can't use (e.g.) `MyService x = new ServiceTracker().open()`

What is the difference to having it in multiple line? (beside it needs "more code")?

> and then deal with any errors that might occur during the call of the open itself.

The only exception open can throw is if the bundle is no longer active in which case you are lost anyways ...

> This method added appears to violate the dynamism of the OSGi API and to return a service which could be inadvertently cached in an instance variable, which would be incorrect.

What method you exactly reference to? Beside that one has to keep up with reality, actually *any* technique can prevent anyone from keeping a reference and even if OSGi is dynamic especially Eclipse is not that dynamic, I try to change this here and there but we still need some intermediate solution to make the people happy that are used to the static approach.

But as no one of those seem to be interested in this, its probably just fear-of-change than an actual problem and we better do nothing at all so closing it for now.

**Comment by tjwatson** [comment]
> The specific problem is that to use a `ServiceTracker` can't be used to interact with services until the `.open()` call is invoked. However, when `.open()` is invoked it triggers a servlce lookup. This in turn is picked up by DS which then eagerly forces the activation of the service **during the `.open()` call synchronously** so that by the time the `.open()` call returns, it has activated all services which matched during the lookup from DS providers which otherwise would be lazily used on demand.

TLDR all the content here.  But this particular paragraph I don't fallow.  `ServiceTracker.open` only looks at the service registry to find the existing `ServiceReference` objects.  This does not in any way trigger DS to start looking at anything that would trigger activation of service components.  It is not until one of the `ServiceTracker.getService` or `ServiceTracker.getServices` methods is called does it actually force the creating of a service component.  Just looking for a `ServiceReference` based on some filter has no effect on the system other than evaluating service properties of the existing service registrations.  That should not result in more services being registered or activated.
